test/integration/scheduler/framework_test.go

```diff
@@ -1020,124 +1020,124 @@ func TestPrebindPlugin(t *testing.T) {
-// 			prof := schedulerconfig.KubeSchedulerProfile{
-// 				SchedulerName: v1.DefaultSchedulerName,
-// 				Plugins: &schedulerconfig.Plugins{
-// 					Reserve: schedulerconfig.PluginSet{
-// 						// filled by looping over reservePlugins
-// 					},
-// 					PreBind: schedulerconfig.PluginSet{
-// 						Enabled: []schedulerconfig.Plugin{
-// 							{
-// 								Name: preBindPluginName,
-// 							},
-// 						},
-// 					},
-// 				},
-// 			}
-// 			for _, pl := range reservePlugins {
-// 				prof.Plugins.Reserve.Enabled = append(prof.Plugins.Reserve.Enabled, schedulerconfig.Plugin{
-// 					Name: pl.Name(),
-// 				})
-// 			}
-
-// 			// Create the master and the scheduler with the test plugin set.
-// 			testCtx := initTestSchedulerForFrameworkTest(t, testutils.InitTestMaster(t, "reserve-plugin-unreserve", nil), 2,
-// 				scheduler.WithProfiles(prof),
-// 				scheduler.WithFrameworkOutOfTreeRegistry(registry))
-// 			defer testutils.CleanupTest(t, testCtx)
-
-// 			preBindPlugin.failPreBind = test.failPreBind
-// 			if test.failReserve {
-// 				reservePlugins[test.failReserveIndex].failReserve = true
-// 			}
-// 			// Create a best effort pod.
-// 			pod, err := createPausePod(testCtx.ClientSet,
-// 				initPausePod(&pausePodConfig{Name: "test-pod", Namespace: testCtx.NS.Name}))
-// 			if err != nil {
-// 				t.Errorf("Error while creating a test pod: %v", err)
-// 			}
-
-// 			if test.failPreBind || test.failReserve {
-// 				if err = wait.Poll(10*time.Millisecond, 30*time.Second, podSchedulingError(testCtx.ClientSet, pod.Namespace, pod.Name)); err != nil {
-// 					t.Errorf("Expected a scheduling error, but didn't get it: %v", err)
-// 				}
-// 				for i := numReservePlugins - 1; i >= 0; i-- {
-// 					select {
-// 					case event := <-pluginInvokeEventChan:
-// 						expectedPluginName := reservePlugins[i].Name()
-// 						if expectedPluginName != event.pluginName {
-// 							t.Errorf("event.pluginName = %s, want %s", event.pluginName, expectedPluginName)
-// 						}
-// 					case <-time.After(time.Second * 30):
-// 						t.Errorf("pluginInvokeEventChan receive timed out")
-// 					}
-// 				}
-// 			} else {
-// 				if err = testutils.WaitForPodToSchedule(testCtx.ClientSet, pod); err != nil {
-// 					t.Errorf("Expected the pod to be scheduled, got an error: %v", err)
-// 				}
-// 				for i, pl := range reservePlugins {
-// 					if pl.numUnreserveCalled != 0 {
-// 						t.Errorf("reservePlugins[%d].numUnreserveCalled = %d, want 0", i, pl.numUnreserveCalled)
-// 					}
-// 				}
-// 			}
-// 			testutils.CleanupPods(testCtx.ClientSet, t, []*v1.Pod{pod})
-// 		})
-// 	}
-// }
+// TestUnReserveMultiReservePlugins tests unreserve of multi reserve plugins.
+func TestUnReserveMultiReservePlugins(t *testing.T) {
+	tests := []struct {
+		name string
+		fail bool
+	}{
+		{
+			name: "success reserve plugin",
+			fail: false,
+		},
+		{
+			name: "fail reserve plugin",
+			fail: true,
+		},
+	}
+
+	for index, test := range tests {
```

Comment:
> Using a channel looks unnecessarily complex to me. I'd suggest instantiating a `*int` counter in each subtest, and pass it to each Reserve plugin. Suppose we have 3 Reserve plugins. We can implement the Reserver() as adding 2^i in each plugin - i.e., each Reserve plugin to add up 2/4/8 to the counter, and implement Unreserve() as deducting 2^i-1 in each plugin.
> 
> So that we can compose the following cases:
> 
> * The first Reserve plugin fails
> * The last Reserve plugin fails
> * The Reserve plugin in the middle fails
> * All Reserve plugins pass, but a Permit plugin fails
> * All Reserve plugins pass, but a Permit plugin times out
> * All Reserve plugins pass, but a PreBind plugin fails
> * All Reserve plugins pass, but a Bind plugin fails

sounds good. Busy working over the past few days.

Reply:
> This applies an extra outer loop to `t.Run()` at L1094, which doesn't look right.

a litter confused about this advice. as testname in `t.Run(testname, func)` is inheritance from the outer loop. However, I'll refactor this as advised above.
```
	for index, test := range tests {}
```

Reply:
Oh, I misread... But you can put the initialization logic into t.Run().

Reply:
got it. :)